dartagnan/src/main/java/com/dat3m/dartagnan/program/event/lang/dat3m/DynamicThreadJoin.java

```diff
@@ -0,0 +1,89 @@
+package com.dat3m.dartagnan.program.event.lang.dat3m;
+
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.ExpressionVisitor;
+import com.dat3m.dartagnan.expression.type.AggregateType;
+import com.dat3m.dartagnan.expression.type.IntegerType;
+import com.dat3m.dartagnan.expression.utils.IntegerHelper;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.event.AbstractEvent;
+import com.dat3m.dartagnan.program.event.BlockingEvent;
+import com.dat3m.dartagnan.program.event.RegReader;
+import com.dat3m.dartagnan.program.event.RegWriter;
+import com.google.common.base.Preconditions;
+
+import java.math.BigInteger;
+import java.util.HashSet;
+import java.util.Set;
+
+
+/*
+    This event joins with given thread, i.e., it blocks until the target thread terminates properly.
+    - The target thread is specified by a thread id, typically the one returned by DynamicThreadCreate.
+    - The type of the return register must be { bv? status, TRet retValue } where TRet may be void if no value is returned.
+      The status may indicate success (status == 0) or an error (status != 0).
+      The status register shall be large enough to accommodate all error values (1 byte should be sufficient).
+    - The return value retValue is only well-defined in the case of success (status == 0).
+    - It is possible to join with the same thread multiple times.
+ */
+public class DynamicThreadJoin extends AbstractEvent implements RegWriter, RegReader, BlockingEvent {
+
+    public enum Status {
+        SUCCESS,                // The join was successful
+        INVALID_TID,            // The provided tid does not match with a joinable thread
+        INVALID_RETURN_TYPE     // The provided tid is valid, but the return type of that thread does not match
+                                // with the expected return type.
+    }
+
+    protected Register resultRegister;
+    protected Expression tid;
+
+    public DynamicThreadJoin(Register resultRegister, Expression tid) {
+        Preconditions.checkArgument(resultRegister.getType() instanceof AggregateType aggType
```

Comment:
why it the return type an aggregate? what are the other two fields used for?

Reply:
You quoted the comment that explains it above :P.
There are only two fields: the first is the status and the second is value returned by the thread.